### ISSUE TITLE
Possible deadlock in [YapDatabaseConnection dealloc]

### ISSUE THREAD

**Issue by Wisors** [yapstudios/YapDatabase#437]
We have a lot of crash reports in our application that looks like a deadlock caused by nested dispatch_sync calls to database snapshotQueue. This is one of them, I pointed nested calls here:

> Thread 13 Crashed:
> 0   libdispatch.dylib                    0x0000000184a4f0d0 _dispatch_sync_wait + 600
> 1   YapDatabase                          0x00000001064d8954 -**[YapDatabaseConnection postReadTransaction:] (YapDatabaseConnection.m:2468)** -> **dispatch_sync(database->snapshotQueue**
> 2   YapDatabase                          0x00000001064d1e48 __32-[YapDatabaseConnection dealloc]_block_invoke (YapDatabaseConnection.m:409)
> 3   libdispatch.dylib                    0x0000000184a41048 _dispatch_client_callout + 12
> 4   libdispatch.dylib                    0x0000000184a49ae8 _dispatch_queue_barrier_sync_invoke_and_complete + 52
> 5   YapDatabase                          0x00000001064d1c2c -[YapDatabaseConnection dealloc] (YapDatabaseConnection.m:417)
> 6   YapDatabase                          0x0000000106496cdc -[YapDatabase noteCommittedChangeset:fromConnection:] (YapDatabase.m:2747)
> 7   YapDatabase                          0x00000001064da5c8 __50-[YapDatabaseConnection postReadWriteTransaction:]_block_invoke.392 (YapDatabaseConnection.m:3025)
> 8   libdispatch.dylib                    0x0000000184a41048 _dispatch_client_callout + 12
> 9   libdispatch.dylib                    0x0000000184a49ae8 _dispatch_queue_barrier_sync_invoke_and_complete + 52
> 10  YapDatabase                          0x00000001064d9d9c -**[YapDatabaseConnection postReadWriteTransaction:] (YapDatabaseConnection.m:3017)** -> **_dispatch_sync(database->snapshotQueue..._**
> 11  YapDatabase                          0x00000001064d7bfc __81-[YapDatabaseConnection asyncReadWriteWithBlock:completionQueue:completionBlock:]_block_invoke_2 (YapDatabaseConnection.m:2157)
> 12  libdispatch.dylib                    0x0000000184a41048 _dispatch_client_callout + 12
> 13  libdispatch.dylib                    0x0000000184a49ae8 _dispatch_queue_barrier_sync_invoke_and_complete + 52
> 14  YapDatabase                          0x00000001064d7b34 __81-[YapDatabaseConnection asyncReadWriteWithBlock:completionQueue:completionBlock:]_block_invoke (YapDatabaseConnection.m:2151)
> 15  libdispatch.dylib                    0x0000000184a41088 _dispatch_call_block_and_release + 20
> 16  libdispatch.dylib                    0x0000000184a41048 _dispatch_client_callout + 12
> 17  libdispatch.dylib                    0x0000000184a4ae48 _dispatch_queue_serial_drain$VARIANT$mp + 524
> 18  libdispatch.dylib                    0x0000000184a4b7d8 _dispatch_queue_invoke$VARIANT$mp + 336
> 19  libdispatch.dylib                    0x0000000184a4c200 _dispatch_root_queue_drain_deferred_wlh$VARIANT$mp + 396
> 20  libdispatch.dylib                    0x0000000184a544a0 _dispatch_workloop_worker_thread$VARIANT$mp + 640
> 21  libsystem_pthread.dylib              0x0000000184ce6fd0 _pthread_wqthread + 928
> 22  libsystem_pthread.dylib              0x0000000184ce6c20 start_wqthread + 0

We have a setup with two connections:
1) connection for synchronous reads(it is calling only synchronous API of YapDatabaseConnection)
2) connection for asynchronous writes (it is calling only asynchronous API of YapDatabaseConnection)

Looks like while we are committing some changes, user dismissed screen and we lost the last strong reference to our connections. 
At the same time, the writable connection propagates changes to other database's connections and uses `database->snapshotQueue` for it.

> -**[YapDatabaseConnection postReadWriteTransaction:] (YapDatabaseConnection.m:3017)** 
  -> `_dispatch_sync(database->snapshotQueue`....
  -> `[YapDatabase noteCommittedChangeset:fromConnection:]`

The database has a loop through connections and it temporary acquires a strong reference to each of them. Here, we have a dealloc call to iterated connection and it leads to one more `_dispatch_sync( database->snapshotQueue,..` call and it causes the deadlock.

What can you suggest here? Is it possible to eliminate this deadlock without changing our approach with multiple connections?

**Comment by Wisors** [comment]
I continued my investigation. I found that we have one more connection in our setup and this one is `longLivedReadTransaction`. Provided deadlock is happening during deallocation of this third connection. So, the final picture here:

1) connection for synchronous reads
2) connection for asynchronous writes
3) `longLivedReadTransaction`

The following happened:

1. We have some async changes inside write connection
2. While it propagates to two others connections we lose external strong reference on the main queue to all of this connections.
3. YapDatabase instance have a temp strong reference to connections inside `[YapDatabase noteCommittedChangeset:fromConnection:]`
4. As soon as this temp reference released we have a `[YapDatabaseConnection dealloc]` call in the same background thread.
5. Because one of our connections is longLivedReadTransaction it posts read transaction to already executing `database->snapshotQueue`.

The documentation said that we don't need to explicitly cleanup longLivedReadTransaction, but for this case, it looks like it is the easiest solution to fix deadlock.
